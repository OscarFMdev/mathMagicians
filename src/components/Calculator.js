import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../logic/calculate';

const Buttons = ({ myFunction }) => (
  <>
    <button onClick={myFunction} type="button" className="blue-button">AC</button>
    <button onClick={myFunction} type="button" className="blue-button">+/-</button>
    <button onClick={myFunction} type="button" className="blue-button">%</button>
    <button onClick={myFunction} type="button" className="orange-button">รท</button>
    <button onClick={myFunction} type="button" className="blue-button">7</button>
    <button onClick={myFunction} type="button" className="blue-button">8</button>
    <button onClick={myFunction} type="button" className="blue-button">9</button>
    <button onClick={myFunction} type="button" className="orange-button">x</button>
    <button onClick={myFunction} type="button" className="blue-button">4</button>
    <button onClick={myFunction} type="button" className="blue-button">5</button>
    <button onClick={myFunction} type="button" className="blue-button">6</button>
    <button onClick={myFunction} type="button" className="orange-button">-</button>
    <button onClick={myFunction} type="button" className="blue-button">1</button>
    <button onClick={myFunction} type="button" className="blue-button">2</button>
    <button onClick={myFunction} type="button" className="blue-button">3</button>
    <button onClick={myFunction} type="button" className="orange-button">+</button>
    <button onClick={myFunction} type="button" className="blue-button large">0</button>
    <button onClick={myFunction} type="button" className="blue-button">.</button>
    <button onClick={myFunction} type="button" className="orange-button">=</button>
  </>
);

const Calculator = () => {
  const [display, setDisplay] = useState(0);
  useEffect(() => {
    setDisplay({
      total: 0,
      next: null,
      operation: null,
    });
  }, []);

  const handleClick = ({ target }) => {
    setDisplay(calculate(display, target.textContent));
  };

  const { total, next, operation } = display;
  return (
    <>
      <section className="calculator-container">
        <div className="calc-quote">Let&apos;s do dome Math!</div>
        <div className="calculator">
          <div className="result-box">
            <p>
              {total}
              {operation}
              {next}
            </p>
          </div>
          <ul className="operation-grid">
            <Buttons myFunction={handleClick} />
          </ul>
        </div>
      </section>
    </>
  );
};

Buttons.propTypes = {
  myFunction: PropTypes.func.isRequired,
};

export default Calculator;
