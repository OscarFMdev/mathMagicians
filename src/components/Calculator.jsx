import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../logic/calculate';

const Buttons = ({ myFunction }) => (
  <>
    <button onClick={myFunction} type="button" className="gray-button">AC</button>
    <button onClick={myFunction} type="button" className="gray-button">+/-</button>
    <button onClick={myFunction} type="button" className="gray-button">%</button>
    <button onClick={myFunction} type="button" className="orange-button">รท</button>
    <button onClick={myFunction} type="button" className="gray-button">7</button>
    <button onClick={myFunction} type="button" className="gray-button">8</button>
    <button onClick={myFunction} type="button" className="gray-button">9</button>
    <button onClick={myFunction} type="button" className="orange-button">x</button>
    <button onClick={myFunction} type="button" className="gray-button">4</button>
    <button onClick={myFunction} type="button" className="gray-button">5</button>
    <button onClick={myFunction} type="button" className="gray-button">6</button>
    <button onClick={myFunction} type="button" className="orange-button">-</button>
    <button onClick={myFunction} type="button" className="gray-button">1</button>
    <button onClick={myFunction} type="button" className="gray-button">2</button>
    <button onClick={myFunction} type="button" className="gray-button">3</button>
    <button onClick={myFunction} type="button" className="orange-button">+</button>
    <button onClick={myFunction} type="button" className="gray-button large">0</button>
    <button onClick={myFunction} type="button" className="gray-button">.</button>
    <button onClick={myFunction} type="button" className="orange-button">=</button>
  </>
);

const Calculator = () => {
  const [display, setDisplay] = useState(0);
  useEffect(() => {
    setDisplay({
      total: 0,
      next: null,
      operation: null,
    });
  }, []);

  const handleClick = (e) => {
    setDisplay(calculate(display, e.target.textContent));
  };

  const { total, next, operation } = display;
  return (
    <section className="calculator-container">
      <div className="calculator">
        <div className="gray-box">
          <p>
            {total}
            {operation}
            {next}
          </p>
        </div>
        <ul className="operation-grid">
          <Buttons myFunction={handleClick} greet="Hello" />
        </ul>
      </div>
    </section>
  );
};

Buttons.propTypes = {
  myFunction: PropTypes.func.isRequired,
};

export default Calculator;
