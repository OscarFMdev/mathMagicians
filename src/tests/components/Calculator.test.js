import { fireEvent, render } from '@testing-library/react';
import Calculator from '../../components/Calculator';

describe('Calculator', () => {
  /* RENDER */
  test('should render', () => {
    const calculator = render(<Calculator />);
    expect(calculator).toMatchSnapshot();
  });

  /* USER INTERACTION */
  test('should modify the result-box content after user pressing the buttons "7"', () => {
    /* Arrange */
    const { container } = render(<Calculator />);
    const buttons = container.querySelectorAll('button');
    const buttonSeven = buttons[4];

    /* Act */
    fireEvent.click(buttonSeven);
    const resultBox = container.querySelector('.result-box > p').innerHTML;

    /* Assert */
    expect(resultBox).toBe('7');
  });
});
