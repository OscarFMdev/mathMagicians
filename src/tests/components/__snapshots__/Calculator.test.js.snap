// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="calculator-container"
      >
        <div
          class="calc-quote"
        >
          Let's do dome Math!
        </div>
        <div
          class="calculator"
        >
          <div
            class="result-box"
          >
            <p>
              0
            </p>
          </div>
          <ul
            class="operation-grid"
          >
            <button
              class="blue-button"
              type="button"
            >
              AC
            </button>
            <button
              class="blue-button"
              type="button"
            >
              +/-
            </button>
            <button
              class="blue-button"
              type="button"
            >
              %
            </button>
            <button
              class="orange-button"
              type="button"
            >
              รท
            </button>
            <button
              class="blue-button"
              type="button"
            >
              7
            </button>
            <button
              class="blue-button"
              type="button"
            >
              8
            </button>
            <button
              class="blue-button"
              type="button"
            >
              9
            </button>
            <button
              class="orange-button"
              type="button"
            >
              x
            </button>
            <button
              class="blue-button"
              type="button"
            >
              4
            </button>
            <button
              class="blue-button"
              type="button"
            >
              5
            </button>
            <button
              class="blue-button"
              type="button"
            >
              6
            </button>
            <button
              class="orange-button"
              type="button"
            >
              -
            </button>
            <button
              class="blue-button"
              type="button"
            >
              1
            </button>
            <button
              class="blue-button"
              type="button"
            >
              2
            </button>
            <button
              class="blue-button"
              type="button"
            >
              3
            </button>
            <button
              class="orange-button"
              type="button"
            >
              +
            </button>
            <button
              class="blue-button large"
              type="button"
            >
              0
            </button>
            <button
              class="blue-button"
              type="button"
            >
              .
            </button>
            <button
              class="orange-button"
              type="button"
            >
              =
            </button>
          </ul>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="calculator-container"
    >
      <div
        class="calc-quote"
      >
        Let's do dome Math!
      </div>
      <div
        class="calculator"
      >
        <div
          class="result-box"
        >
          <p>
            0
          </p>
        </div>
        <ul
          class="operation-grid"
        >
          <button
            class="blue-button"
            type="button"
          >
            AC
          </button>
          <button
            class="blue-button"
            type="button"
          >
            +/-
          </button>
          <button
            class="blue-button"
            type="button"
          >
            %
          </button>
          <button
            class="orange-button"
            type="button"
          >
            รท
          </button>
          <button
            class="blue-button"
            type="button"
          >
            7
          </button>
          <button
            class="blue-button"
            type="button"
          >
            8
          </button>
          <button
            class="blue-button"
            type="button"
          >
            9
          </button>
          <button
            class="orange-button"
            type="button"
          >
            x
          </button>
          <button
            class="blue-button"
            type="button"
          >
            4
          </button>
          <button
            class="blue-button"
            type="button"
          >
            5
          </button>
          <button
            class="blue-button"
            type="button"
          >
            6
          </button>
          <button
            class="orange-button"
            type="button"
          >
            -
          </button>
          <button
            class="blue-button"
            type="button"
          >
            1
          </button>
          <button
            class="blue-button"
            type="button"
          >
            2
          </button>
          <button
            class="blue-button"
            type="button"
          >
            3
          </button>
          <button
            class="orange-button"
            type="button"
          >
            +
          </button>
          <button
            class="blue-button large"
            type="button"
          >
            0
          </button>
          <button
            class="blue-button"
            type="button"
          >
            .
          </button>
          <button
            class="orange-button"
            type="button"
          >
            =
          </button>
        </ul>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
